Sparse CSR: add test parametrization for enabling sparse_csr

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #61858
* #61857
* #65606
* #64391
* #63948
* #63511
* #63510
* #63509
* #61536
* __->__ #61408
* #61407

This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @mruberry

IvanYashchuk committed:
Sparse CSR: add test parametrization for enabling sparse_csr

This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,63 @@ def test_wrapper(*args, **kwargs):
 
+# Decorator that generates two set of tests with kwarg sparse_csr=True and sparse_csr=False.
+# All keyword arguments are used only if sparse_csr==True, so they have no effect on the original test
+class enableSparseCSR(_TestParametrizer):
```
@pmeier, could you please review this PR, particularly `enableSparseCSR` functionality? This adds a simple `sparse_csr=True/False` and a bit of specific customization when `sparse_csr=True`.

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
+# Decorator that generates two set of tests with layout=torch.strided and layout=specified_layout.
+# All keyword arguments are used only for the specified layout, so they have no effect on the original test.
+class enableLayout(_TestParametrizer):
+    def __init__(self, layout: torch.layout, *,
```
1. Would it make sense to allow an arbitrary number of layout?  Meaning I could do `@enableLayout(torch.sparse_coo, sparse_scr)` to get 3 tests.
2. If we do 1., does it make sense to always use strided or are there use cases where I for example want sparse COO and CSR but no strided tensors?

IvanYashchuk replied:
1. Yes, I'd modify the code later when such use case arises. Not doing it now because the dtype, device, and operator coverage is not the same for CSR and COO matrices, so separate dtypes and device skips would be needed for each layout.

I think it would better to have in the future the possibility to compose the decorators
```python
@enableLayout(torch.sparse_coo, dtypes=..., )
@enableLayout(torch.sparse_csr, dtypes=..., )
```
to get 3 tests.
2. There would only be only a few functions specific to the sparse layout, so I think using always strided as default layout is fine.

pmeier commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
+                 only_cuda: bool = False,
+                 skip_cuda_if_no_cusparse: bool = True):
+        if not isinstance(layout, torch.layout):
+            raise RuntimeError("Expected to get torch.layout as layout, but got ", layout)
```
NIt: 

1. Since we are checking the type here, it would make sense to also report the type back in the message
2. f-strings :stuck_out_tongue: 

```suggestion
            raise RuntimeError(f"Expected to get torch.layout as layout, but got {type(layout)} instead.")
```

pmeier commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
+        self.dtypes = set(dtypes) if dtypes is not None else None
+        self.dtypes_cpu = set(dtypes_cpu) if dtypes_cpu is not None else None
+        self.dtypes_cuda = set(dtypes_cuda) if dtypes_cuda is not None else None
+        self.only_cuda = only_cuda
```
What is the use case for not doing the test on CPU?

IvanYashchuk replied:
When the tested function is implemented only for cuda and not for cpu. When I have the opposite situation I'd add `only_cpu`.

pmeier replied:
Excuse my ignorance here, but do these cases actually exist? Do we have operators that only work on CUDA? Or do you mean a test that specifically deals with some CUDA behavior?

Regardless, can't we use the [`@onlyCPU` and `@onlyCUDA` decorators](https://github.com/pytorch/pytorch/blob/9b908ab0d0a947d89ac3137f8c4a05a87c35f568/torch/testing/_internal/common_device_type.py#L1076-L1081) to get what we need?

IvanYashchuk replied:
During the development, it's rarely the case to implement the CUDA and CPU support simultaneously. For example in the current stack of PRs, first CUDA support is added, then CPU. So the operator can work only for CUDA, or only for CPU, or only for specific dtypes.
`onlyCUDA` decorator is used here if `only_cuda=True`: https://github.com/pytorch/pytorch/blob/d9de883b87ea39054bb99986775fc76b6fc54ece/torch/testing/_internal/common_device_type.py#L843-L844
dtype/device coverage of sparse_csr layout is very different from strided layout, that's why we need to keep these granular skips separate for these layouts. Instead of adding `@onlyCUDAIfSparseCSR` I added the flag here.

pmeier replied:
> `onlyCUDA` decorator is used here if `only_cuda=True`

Either I still don't fully understand its purpose, but isn't 

```python
@enableSparseCSR(..., only_cuda=True)
def test_foo(self, device, dtype, layout):
    pass
```

equivalent to

```python
@onlyCUDA
@enableSparseCSR(...)
def test_foo(self, device, dtype, layout):
    pass
```

? If yes, I'm against having the `only_cuda` kwarg, because this functionality already exists in an convenient way.

> type/device coverage of sparse_csr layout is very different from strided layout, that's why we need to keep these granular skips separate for these layouts.

We have the `dtypes_cpu` and `dtypes_cuda` kwargs and I'm all for adding them. I currently don't see the connection to `only_cuda` though.

pmeier replied:
Ok, I think I got it now. The use case is that you have an existing operator that already has CPU / CUDA support for strided tensors and you want to add CPU **or** CUDA sparse support for this operator, correct? If yes, sorry for the confusion.

---

Would it make sense to make this a little more general similar to the logic in `instantiate_device_type_tests`? Instead of having the `only_cuda` kwarg, have a `except_for` and `only_for` pair and use every device by default?

pmeier commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
+    def _parametrize_test(self, test, generic_cls, device_cls):
+        """ Parameterizes the given test function with boolean sparse_csr flag. """
+        # if self.layout == torch.strided generate only one set of tests
+        maybe_layout = (self.layout,) if self.layout != torch.strided else tuple()
```
Why would I use the decorator if I only want strided tests?

IvanYashchuk replied:
For no reason, the options for `enableLayout(torch.strided)` are:
1. do nothing special, just like the decorator wasn't there,
2. raise an error
I went with the first option.

pmeier replied:
For me using `@enableLayout(torch.strided)` feels weird, because I would expect that be the default anyway. Without background information I would question why this decorator is there and if it does anything else. Since it is actually a no-op, I would prefer to raise an error and in turn to not allow the decorator in such a case.

jbschlosser replied:
Is this `maybe_layout` logic still needed now that the above has been updated to error out when `layout=torch.strided` is passed in?

IvanYashchuk replied:
No, I should remove that. Thanks!

pmeier commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
+                                               _dtype_test_suffix(dtype))
+
+                # Construct parameter kwargs to pass to the test
+                param_kwargs = {'layout': layout}
+                _update_param_kwargs(param_kwargs, 'dtype', dtype)
```
Do we also need to pass `device` here or is this done by the instantiation function?

IvanYashchuk replied:
This is done by the instantiation function.

pmeier commented on test/test_linalg.py:
```diff
@@ -6014,19 +6043,23 @@ def maybe_transpose(cond, m):
     @tf32_on_and_off(0.005)
-    def test_addmm_sizes(self, device, dtype):
+    @enableSparseCSR(only_cuda=True,
+                     dtypes_cuda=(*torch.testing.get_all_complex_dtypes(),
```
I see there is a `@dtypesIfCUDA` decorator and the `dtypes_cuda` flag. How does the interplay work?

IvanYashchuk replied:
`@dtypesIfCUDA` applies only to normal test and `@enableSparseCSR(dtypes_cuda=...)` applies to sparse_csr layout tests. If `dtypes_cuda` is omitted then dtypes specified by `@dtypesIfCUDA, @dtypesIfCPU, @dtypes` are used for both sets of tests.

pmeier commented on test/test_linalg.py:
```diff
@@ -5901,11 +5908,22 @@ def _test_addmm_addmv(self, f, t, m, v, *, alpha=None, beta=None, transpose_out=
+        if layout == torch.sparse_csr and m.dtype == torch.float16:
+            self.precision = 1e-1
+
+        def to_sparse_csr(mat):
```
This is a misnomer, because it only converts to `sparse_csr` if the respective layout is used. Maybe something like `convert_layout`?

pmeier reviewed (commented):
Hey @IvanYashchuk thanks for the PR. I have a few comments inline.

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

mruberry commented:
Hey @IvanYashchuk, @pmeier! @jbschlosser is looking at some updates to test parameterization. I think it'd be useful to get some face time with him to discuss the best way to parametrize over layouts. I'll ping you on Slack to setup a call.

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -788,6 +788,74 @@ def test_wrapper(*args, **kwargs):
 
+# Decorator that generates two set of tests with layout=torch.strided and layout=specified_layout.
+# All keyword arguments are used only for the specified layout, so they have no effect on the original test.
+class enableLayout(_TestParametrizer):
```
Suggested alternative name: addLayout

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

jbschlosser reviewed (commented):
Looks awesome! One minor comment below

mruberry commented:
Just remember to ping me and @jbschlosser for a final review once the tests are passing and you're satisfied with this PR, @IvanYashchuk

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -775,6 +775,72 @@ def test_wrapper(*args, **kwargs):
+                _update_param_kwargs(param_kwargs, 'dtype', dtype)
+
+                active_decorators = []
+                if non_default_layout and self.skip_cuda_if_no_cusparse:
```
Whew, good. Thanks for the clarification, @IvanYashchuk

jbschlosser commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -775,6 +775,72 @@ def test_wrapper(*args, **kwargs):
+            for dtype in dtypes:
+                layout_str = str(layout).replace('torch.', '_')
+                # Construct the test name
+                test_name = '{}{}_{}{}'.format(test.__name__,
```
Note: The generic test parametrization work has landed, and it included a tweak to how test names should be specified for `_TestParametrizer` subclasses. No need to include `test.__name__` here now; that responsibility is handled higher up.

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk commented:
> Just remember to ping me and @jbschlosser for a final review once the tests are passing and you're satisfied with this PR, @IvanYashchuk

All CI is green and Windows is being tested, @mruberry and @jbschlosser could you please take another look?

cpuhrsch commented on test/test_linalg.py:
```diff
@@ -5936,11 +5943,22 @@ def _test_addmm_addmv(self, f, t, m, v, *, alpha=None, beta=None, transpose_out=
-        res1 = f(t, m, v, alpha=alpha, beta=beta)
+
+        # float16 inputs with cusparse backend require larger tolerance than with cublas
+        if layout == torch.sparse_csr and m.dtype == torch.float16:
```
This seems fishy. Do you know why?

IvanYashchuk replied:
I don't know the reason other than cusparse performs multiplications and accumulation in different order.

cpuhrsch replied:
Maybe cc @xwang233

cpuhrsch commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -775,6 +775,72 @@ def test_wrapper(*args, **kwargs):
 
+# Decorator that generates two set of tests with layout=torch.strided and layout=specified_layout.
+# All keyword arguments are used only for the specified layout, so they have no effect on the original test.
+class addLayout(_TestParametrizer):
```
While I'm a fan of more generic operators, I'd suggest we expand this as we need to support more layouts, especially if implemented in such a central place. This should simplify the implementation and make this easier to merge.

IvanYashchuk replied:
Are you suggesting changing the current code:
```python
class addLayout(_TestParametrizer):
    def __init__(self, layout: torch.layout): pass
addLayoutSparseCSR = partial(addLayout, torch.sparse_csr)
```
to 
```python
class addLayoutSparseCSR(_TestParametrizer):
    def __init__(self, layout: torch.layout):
        self.layout = torch.sparse_csr
        pass
```
? Would it be easier to merge if `addLayout` was prepended with an underscore: `addLayout` -> `_addLayout`? The whole file is in the `_internal` module.

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: add test parametrization for enabling sparse_csr"


This PR adds a new subclass of `_TestParametrizer` that would allow
reusing existing tests by parametrizing whether sparse_csr variant of
the test should be generated or not. This is done by adding a boolean
kwarg `sparse_csr` to the test, which can be used to generate sparse
inputs for the tested functions.
The generated test has name: test_name + 'sparse_csr' + device + dtype.
When `sparse_csr==False` the original test is generated with no
modifications.
`enableSparseCSR` allows specifying CSR specific dtypes and device
skips. Once CSR matrices are better supported the dtype and device
coverage would be the same as for dense tensors.

cc nikitaved pearu cpuhrsch @IvanYashchuk mruberry

[ghstack-poisoned]

IvanYashchuk commented:
It was decided not to add the layout parameterization decorator at this time (https://github.com/pytorch/pytorch/pull/61408#discussion_r726635586). I'll close this PR and submit a new one with normal tests added to `test_sparse_csr.py`.
